plugins/inventory/d42.py

```diff
@@ -115,3 +114,537 @@ def parse(self, inventory, loader, path, cache=False):
+                 WHERE pm.type_id = 3 AND pm.hdsize IS NOT NULL AND p.pcount > 0 AND p.device_fk IS NOT NULL
+                 GROUP BY p.device_fk
+             ) AS p ON view_device_v1.device_pk = p.device_fk
+        """
+
+        return self.get_doql_csv(device_query)
+
+
+    def get_custom_fields(self):
+
+        custom_field_query = """
+                select
+                view_device_v1.device_pk,
+                view_device_v1.name,
+                custom_field.key as custom_field_key,
+                custom_field.value as custom_field_value,
+                custom_field.notes as custom_field_notes
+                from view_device_v1
+                inner join (select device_fk, type, notes, key,
+                                  (CASE
+                                    WHEN view_device_custom_fields_v1.type = 'Related Field'
+                                        THEN (CASE
+                                                WHEN view_device_custom_fields_v1.related_model_name = 'endusers'
+                                                THEN (SELECT name FROM view_enduser_v1 WHERE view_enduser_v1.enduser_pk = view_device_custom_fields_v1.value :: int)
+                                                ELSE 'not available'
+                                                END)
+                                    ELSE view_device_custom_fields_v1.value
+                                    END) as value
+                from view_device_custom_fields_v1) custom_field
+                on view_device_v1.device_pk = custom_field.device_fk
+            """
+
+        return self.get_doql_csv(custom_field_query)
+
+
+    def get_external_links(self):
+        external_links_query = """
+                select
+                view_device_v1.device_pk,
+                view_device_v1.name as device_name,
+                external_links.device_url,
+                external_links.device_url_notes
+                from view_device_v1
+                inner join (select device_fk, device_url, notes as device_url_notes from view_deviceurl_v1) external_links
+                on view_device_v1.device_pk=external_links.device_fk
+            """
+        return self.get_doql_csv(external_links_query)
+
+
+    def get_hdd_details(self):
+        hdd_details_query = """
+        select
+        view_device_v1.device_pk,
+        view_device_v1.name as device_name,
+        hdd_details.raid_group,
+        hdd_details.hdd_serial_no,
+        hdd_details.hdd_raid_type,
+        hdd.hdd_description,
+        hdd.partno,
+        hdd.rpm as hdd_rpm,
+        hdd.rpm_id as hdd_rom_id,
+        hdd.notes as hdd_notes,
+        hdd.hdd_bytes,
+        hdd.location as hdd_location,
+        hdd.hdd_id,
+        hdd.manufactuerer_id,
+        hdd.hdd_type_id,
+        hdd.hdd_type_name,
+        hdd.size,
+        hdd_details.part_id,
+        hdd_details.hddcount,
+        hdd_details.description
+        from view_device_v1
```

Comment:
Do you even need the join on view_device_v1 since view_part_v1 has the device_fk?

Reply:
> I made all the queries perform a join strictly for device name. the name is not used when creating the json, but I wanted the queries to be reusable and be useful for debugging. if for any reason a device is not being imported support can call the same query and find what device information is not coming in for what device